c/tskit/tables.c

```diff
@@ -8623,12 +8732,34 @@ tsk_table_collection_sort(
     tsk_table_collection_t *self, const tsk_bookmark_t *start, tsk_flags_t options)
 {
     int ret = 0;
+    tsk_id_t k;
+    bool canonical = !!(options & TSK_SORT_CANONICAL);
+    tsk_flags_t subset_options = options & TSK_UNCHANGED_POPULATIONS;
+    tsk_id_t *nodes;
     tsk_table_sorter_t sorter;
 
     ret = tsk_table_sorter_init(&sorter, self, options);
     if (ret != 0) {
         goto out;
     }
+
+    if (canonical) {
+        nodes = malloc(self->nodes.num_rows * sizeof(*nodes));
+        if (nodes == NULL) {
+            ret = TSK_ERR_NO_MEMORY;
+            goto out;
+        }
+        for (k = 0; k < (tsk_id_t) self->nodes.num_rows; k++) {
+            nodes[k] = k;
+        }
+        subset_options
+            |= (TSK_SUBSET_KEEP_UNUSED_POPULATIONS | TSK_SUBSET_KEEP_UNUSED_INDIVIDUALS
+                | TSK_SUBSET_KEEP_UNUSED_SITES);
```

Comment:
I am tempted to make a single macro that is `TSK_SUBSET_KEEP_UNUSED = TSK_SUBSET_KEEP_UNUSED_POPULATIONS | TSK_SUBSET_KEEP_UNUSED_INDIVIDUALS
                | TSK_SUBSET_KEEP_UNUSED_SITES`.

Reply:
however, I think someone's going to want to use only some of these options at some point, so I vote to keep them separate.